### PR TITLE
Implements IPMI server

### PR DESCRIPTION
**Pull request by ysksuzuki**
This PR implements IPMI server for virtual BMC, which supports

- IPMI v2.0/RMCP+ authentication and session management
  - Only supports default algorithm, authentication algorithm HMAC-SHA1, integrity algorithm HMAC-SHA1-96, and confidentiality algorithm AES-CBC-128 
- Operations
  - Chassis Power On
  - Chassis Power Off
  - Chassis Power Reset/Cycle

### CODE REVIEW

File: v2/pkg/virtualbmc/bmc_user.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	log.Info("BMC USer: Add user", map[string]interface{}{"user": name})
+}
+
+// RemoveBMCUser removes the user specified from the holder
+func (b *BMCUserHolder) RemoveBMCUser(name string) {
+	_, ok := b.users[name]
+
+	if ok {
+		delete(b.users, name)
+	}
```
Comment (ymmt2005) [inline v2/pkg/virtualbmc/bmc_user.go:39]:
```suggestion
	delete(b.users, name)
```

File: v2/pkg/virtualbmc/bmc_user.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+// GetBMCUser get the user specified from the holder
+func (b *BMCUserHolder) GetBMCUser(name string) (*BMCUser, bool) {
+	obj, ok := b.users[name]
+
+	return obj, ok
```
Comment (ymmt2005) [inline v2/pkg/virtualbmc/bmc_user.go:46]:
```suggestion
	return b.users[name]
```

File: v2/pkg/virtualbmc/ipmi.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+const (
+	CompletionCodeOK              = 0x00
+	CompletionCodeInvalidUsername = 0x81
+)
+
+// IPMI represents IPMIMessage and a target VM
+type IPMI struct {
+	message *IPMIMessage
+	vm      VM
```
Comment (ymmt2005) [inline v2/pkg/virtualbmc/ipmi.go:37]: I'd like to suggest `machine Machine` instead of VM.

File: v2/pkg/virtualbmc/ipmi_server.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+// NewIPMIServer creates an IPMIServer
+func NewIPMIServer(bridges []*Bridge) (*IPMIServer, error) {
+	return &IPMIServer{
+		bridges: bridges,
+	}, nil
+}
+
+func (s *IPMIServer) listen(ctx context.Context, addr string, port int, vm VM) error {
```
Comment (ymmt2005) [inline v2/pkg/virtualbmc/ipmi_server.go:39]:
```suggestion
func (s *IPMIServer) start(ctx context.Context, conn net.PacketConn, vm VM) error {
```

File: v2/pkg/virtualbmc/ipmi_server.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	"net"
+	"strconv"
+
+	"github.com/cybozu-go/log"
+	"github.com/vishvananda/netlink"
+)
+
+// IPMIServer represents IPMI Server
+type IPMIServer struct {
+	bridges []*Bridge
```
Comment (ymmt2005) [inline v2/pkg/virtualbmc/ipmi_server.go:16]: link netlink.Link
```suggestion
	link netlink.Link
```

File: v2/pkg/virtualbmc/ipmi_server.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/cybozu-go/log"
+	"github.com/vishvananda/netlink"
+)
+
+// IPMIServer represents IPMI Server
+type IPMIServer struct {
```
Comment (ymmt2005) [inline v2/pkg/virtualbmc/ipmi_server.go:15]: We will later add Redfish.  Please consider a more appropriate name.

File: v2/pkg/virtualbmc/rmcp_plus.go [hunk 7]
```diff
@@ -0,0 +1,711 @@
...
+	padSize := aes.BlockSize - (len(data) % aes.BlockSize)
+	pad := make([]byte, padSize)
+	for i := range pad {
+		pad[i] = byte(i + 1)
+	}
+	pad[padSize-1] = byte(padSize - 1)
+	return append(data, pad...)
+}
+
+const RandomNumberSize = 16
```
Comment (ymmt2005) [inline v2/pkg/virtualbmc/rmcp_plus.go:694]: No need to be exported.

File: v2/pkg/virtualbmc/ipmi_server.go (hunk 6 above)
Comment (ysksuzuki) [reply v2/pkg/virtualbmc/ipmi_server.go]: Renamed to `BMCServer`

File: v2/pkg/virtualbmc/ipmi_server.go (hunk 5 above)
Comment (ysksuzuki) [reply v2/pkg/virtualbmc/ipmi_server.go]: Removed network handlings from BMCServer and leave it to a client that uses BMCServer.

File: v2/pkg/virtualbmc/bmc_user.go (hunk 2 above)
Comment (ysksuzuki) [reply v2/pkg/virtualbmc/bmc_user.go:42]: It seems that the result of the test whether a key is present needs to be received with two-value assignment to return it from a function. Got the error "not enough arguments to return" when I returned two values directly.

Comment (ymmt2005) [reply v2/pkg/virtualbmc/bmc_user.go:42]: Oh, I didn't know that.

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -8,11 +8,11 @@ defaults:
   run:
     working-directory: v2
 env:
-  go-version: 1.13
+  go-version: 1.15
 jobs:
   test:
     name: Small test
-    runs-on: ubuntu-18.04
```
Comment (ymmt2005) [inline .github/workflows/ci.yaml:15]: Could you do a matrix test for 18.04 and 20.04?

File: v2/pkg/virtualbmc/bmc_server.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+type Machine interface {
+	IsRunning() bool
+	PowerOn() error
+	PowerOff() error
+}
+
+// NewBMCServer creates an BMCServer
+func NewBMCServer() (*BMCServer, error) {
+	return &BMCServer{}, nil
+}
```
Comment (ymmt2005) [inline v2/pkg/virtualbmc/bmc_server.go:26]: This seems unnecessary...

File: v2/pkg/virtualbmc/bmc_server.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+		server.Close()
+	}()
+
+	session := NewRMCPPlusSessionHolder()
+	bmcUser := NewBMCUserHolder()
+	bmcUser.AddBMCUser("cybozu", "cybozu")
+
+	buf := make([]byte, 1024)
+	for {
+		_, addr, err := server.ReadFromUDP(buf)
```
Comment (ymmt2005) [inline v2/pkg/virtualbmc/bmc_server.go:49]: Is it possible to use a more generic `ReadFrom` paired with `WriteTo`?

**Review by ymmt2005 — CHANGES_REQUESTED** [review]

File: v2/pkg/virtualbmc/bmc_server.go (hunk 10 above)
Comment (ysksuzuki) [reply v2/pkg/virtualbmc/bmc_server.go:49]: Do you mean like this?


Server
```go
func Listen(ctx context.Context, conn net.PacketConn, machine Machine) error {
	go func() {
		<-ctx.Done()
		conn.Close()
	}()

	session := NewRMCPPlusSessionHolder()
	bmcUser := NewBMCUserHolder()
	bmcUser.AddBMCUser("cybozu", "cybozu")

	buf := make([]byte, 1024)
	for {
		_, addr, err := conn.ReadFrom(buf)
		if err != nil {
			return err
		}

		bytebuf := bytes.NewBuffer(buf)
		res, err := HandleRMCPRequest(bytebuf, machine, session, bmcUser)
		if err != nil {
			log.Warn("failed to handle RMCP request", map[string]interface{}{
				log.FnError: err,
			})
			continue
		}
		_, err = conn.WriteTo(res, addr)
		if err != nil {
			log.Warn("failed to write to UDP", map[string]interface{}{
				log.FnError: err,
			})
			continue
		}
	}
}
```

Client
```go
var _ = Describe("Virtual BMC", func() {
	It("should turn on and off Machine power via IPMI v2.0", func() {
		serverAddr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", "127.0.0.1", 9623))
		Expect(err).NotTo(HaveOccurred())
		conn, err := net.ListenUDP("udp", serverAddr)
		Expect(err).NotTo(HaveOccurred())
		well.Go(func(ctx context.Context) error {
			return Listen(ctx, conn, &VMMock{running: false})
		})
```

Comment (ymmt2005) [reply v2/pkg/virtualbmc/bmc_server.go:49]: Right.

**Review by ymmt2005 — APPROVED** [review]
LGTM

**Pull request merged**